File: common/src/main/java/dev/mayaqq/estrogen/registry/recipes/ThighHighDyeRecipe.java
package dev.mayaqq.estrogen.registry.recipes;

import com.simibubi.create.foundation.utility.Color;
import dev.mayaqq.estrogen.registry.EstrogenItems;
import dev.mayaqq.estrogen.registry.EstrogenRecipes;
import dev.mayaqq.estrogen.registry.items.ThighHighsItem;
import net.minecraft.core.RegistryAccess;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.inventory.CraftingContainer;
import net.minecraft.world.item.DyeItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.CraftingBookCategory;
import net.minecraft.world.item.crafting.CustomRecipe;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.level.Level;

public class ThighHighDyeRecipe extends CustomRecipe {
    public ThighHighDyeRecipe(ResourceLocation id, CraftingBookCategory category) {
        super(id, category);
    }

    @Override
    public boolean matches(CraftingContainer inv, Level level) {
        for (int i = 0; i < 3; i++) {
            int leftSlot = i * 3;
            int middleSlot = leftSlot + 1;
            int rightSlot = leftSlot + 2;
            if (inv.getItem(middleSlot).getItem() instanceof ThighHighsItem) {
                ItemStack rightToItem = inv.getItem(rightSlot);
                ItemStack leftToItem = inv.getItem(leftSlot);

                if (rightToItem.getItem() instanceof DyeItem || leftToItem.getItem() instanceof DyeItem) {
                    for (int j = 0; j < 9; j++) {
                        if (j == leftSlot || j == middleSlot || j == rightSlot) {
                            continue;
                        }
                        if (!inv.getItem(j).isEmpty()) {
                            return false;
                        }
                    }

                    return EstrogenItems.THIGH_HIGHS.get().hasCustomColor(inv.getItem(middleSlot)) ||
                        (rightToItem.getItem() instanceof DyeItem && leftToItem.getItem() instanceof DyeItem);

                }
            }
        }
        return false;
    }

    @Override
    public ItemStack assemble(CraftingContainer inv, RegistryAccess registryAccess) {
        for (int i = 0; i < 3; i++) {
            int leftSlot = i * 3;
            int middleSlot = leftSlot + 1;
            int rightSlot = leftSlot + 2;
            if (inv.getItem(middleSlot).getItem() instanceof ThighHighsItem thighHighsItem) {
                ItemStack item = inv.getItem(middleSlot);
                ItemStack leftToItem = inv.getItem(leftSlot);
                ItemStack rightToItem = inv.getItem(rightSlot);

                Color newPrimary, newSecondary;

                if(thighHighsItem.hasCustomColor(item)) {
                    Color oldPrimary = new Color(thighHighsItem.getColor(item, 0));
                    Color oldSecondary = new Color(thighHighsItem.getColor(item, 1));
                    newPrimary = mixColorWithDye(oldPrimary, leftToItem);
                    newSecondary = mixColorWithDye(oldSecondary, rightToItem);
                } else {
                    newPrimary = colorFromDye(leftToItem);
                    newSecondary = colorFromDye(rightToItem);
                }

                ItemStack newThighHighsItem = item.copy();
                thighHighsItem.setColor(newThighHighsItem, newPrimary.getRGB(), newSecondary.getRGB());
                return newThighHighsItem;
            }
        }
        return ItemStack.EMPTY;
    }

    private Color mixColorWithDye(Color original, ItemStack dyeStack) {
        if(dyeStack.isEmpty() || !(dyeStack.getItem() instanceof DyeItem)) return original;
        Color dyeColor = colorFromDye(dyeStack);
        return original.mixWith(dyeColor, .5f);
    }

    private Color colorFromDye(ItemStack dyeStack) {
        float[] dyeColors = ((DyeItem) dyeStack.getItem()).getDyeColor().getTextureDiffuseColors();
        return new Color(dyeColors[0], dyeColors[1], dyeColors[2], 1f);
    }

    @Override
    public boolean canCraftInDimensions(int width, int height) {
        return width * height == 9;
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return EstrogenRecipes.THIGH_HIGH_DYE_SERIALIZER.get();
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/content/recipes/ThighHighDyeRecipe.kt
package dev.mayaqq.estrogen.content.recipes

import dev.mayaqq.estrogen.content.EstrogenItems
import dev.mayaqq.estrogen.content.EstrogenRecipeSerializers
import dev.mayaqq.estrogen.content.items.ThighHighsItem
import invoke.kitty.kritter.utils.color.Color
import invoke.kitty.kritter.utils.color.asColor
import invoke.kitty.kritter.utils.extensions.isOf
import net.minecraft.core.HolderLookup
import net.minecraft.world.item.DyeItem
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.CraftingBookCategory
import net.minecraft.world.item.crafting.CraftingInput
import net.minecraft.world.item.crafting.CustomRecipe
import net.minecraft.world.item.crafting.RecipeSerializer
import net.minecraft.world.level.Level


class ThighHighDyeRecipe(category: CraftingBookCategory) : CustomRecipe(category) {
    override fun matches(inv: CraftingInput, level: Level): Boolean {
        inv.columns().let { columns ->
            if (columns.size < 2) return false

            val thighHighs = columns[1].filter { it isOf EstrogenItems.ThighHighs }.apply {
                if (this.size != 1) return false
            }.first()

            val thighHighsItem = thighHighs.item as ThighHighsItem

            if (thighHighsItem.getStyle(thighHighs) != null) return false

            val hasLeftDyes = columns[0].any { it.item is DyeItem }
            val hasRightDyes = columns.getOrNull(2)?.any { it.item is DyeItem } == true

            return if (thighHighsItem.hasCustomColor(thighHighs)) hasLeftDyes || hasRightDyes else hasLeftDyes && hasRightDyes
        }
    }

    override fun assemble(inv: CraftingInput, lookup: HolderLookup.Provider): ItemStack {
        inv.columns().let { columns ->
            val stack = columns[1].first { it isOf EstrogenItems.ThighHighs }
            val item = stack.item as ThighHighsItem
            val leftDye = columns[0].filter { it.item is DyeItem }
            val rightDye = columns.getOrNull(2)?.filter { it.item is DyeItem }?: listOf()

            var newPrimary: Color? = null
            var newSecondary: Color? = null

            if (item.hasCustomColor(stack)) {
                var oldPrimary = item.getColor(stack, 0)
                var oldSecondary = item.getColor(stack, 1)
                leftDye.forEach { dye ->
                    oldPrimary = mixColorWithDye(oldPrimary, dye)
                }
                rightDye.forEach { dye ->
                    oldSecondary = mixColorWithDye(oldSecondary, dye)
                }
                newPrimary = oldPrimary
                newSecondary = oldSecondary
            } else {
                leftDye.forEach { dye ->
                    newPrimary = if (newPrimary == null) colorFromDye(dye) else {
                        mixColorWithDye(newPrimary, dye)
                    }
                }
                rightDye.forEach { dye ->
                    newSecondary = if (newSecondary == null) colorFromDye(dye) else {
                        mixColorWithDye(newSecondary, dye)
                    }
                }
            }

            val newThighHighs = stack.copy()
            item.setColor(newThighHighs, newPrimary!!, newSecondary!!)
            return newThighHighs
        }
    }

    private fun mixColorWithDye(original: Color, dyeStack: ItemStack): Color {
        if (dyeStack.isEmpty || dyeStack.item !is DyeItem) return original
        val dyeColor: Color = colorFromDye(dyeStack)
        return original.mix(dyeColor, .5f)
    }

    private fun colorFromDye(dyeStack: ItemStack): Color {
        return (dyeStack.item as DyeItem).dyeColor.asColor()
    }

    override fun canCraftInDimensions(width: Int, height: Int): Boolean = width * height == 9
    override fun getSerializer(): RecipeSerializer<*> = EstrogenRecipeSerializers.THIGH_HIGH_DYE_SERIALIZER.value!!

    private fun CraftingInput.columns(): List<List<ItemStack>> {
        return this.items().withIndex().groupBy { it.index % 3 }.map { entry -> entry.value.map { it.value } }
    }
}